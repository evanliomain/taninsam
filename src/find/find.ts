/**
 * @module array=>any
 */
/**
 * Returns the value of the first element in the array where predicate is true,
 * and undefined otherwise.
 * @param predicate find calls predicate once for each element of the array,
 * in ascending order, until it finds one where predicate returns true.
 * If such an element is found, find immediately returns that element value.
 * Otherwise, find returns undefined.
 * @return A find function ready to applied on an array
 * @example
 * ```
 * const input = [
 *   { a: 1, b: 'a' },
 *   { a: 2, b: 'b' },
 *   { a: 3, b: 'c' },
 *   { a: 4, b: 'd' }
 * ];
 * find<{ a: number; b: string }[]>(({ a }) => 3 === a)(input)
 * // { a: 3, b: 'c' }
 * ```
 * @example Using the chain
 * ```
 * chain(input)
 *   .chain(find<{ a: number; b: string }[]>(({ a }) => 3 === a))
 *   .value() // { a: 3, b: 'c' }
 * ```
 */
export function find<T>(
  predicate: (value: T, index: number, array: ReadonlyArray<T>) => boolean,
): (array: ReadonlyArray<T>) => T | undefined {
  return (array: ReadonlyArray<T>) => array.find(predicate);
}
