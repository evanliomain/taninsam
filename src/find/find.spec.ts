import { describe, expect, test } from 'vitest';
import { find } from './find';

describe('find function', () => {
  const input = [
    { a: 1, b: 'a' },
    { a: 2, b: 'b' },
    { a: 3, b: 'c' },
    { a: 4, b: 'd' },
  ];

  test('is a pure function', () => {
    find<{ a: number }>(({ a }) => 3 === a)(input);
    expect(input).toMatchSnapshot();
  });

  test('[{ a, b }] |> find(e => e.a === 3) === { a: 3, b: "c" }', () => {
    expect(find<{ a: number }>(({ a }) => 3 === a)(input)).toMatchSnapshot();
  });

  test('[{ a, b }] |> find(e => e.a === 5) === undefined', () => {
    expect(find<{ a: number }>(({ a }) => 5 === a)(input)).toBeUndefined();
  });
});
