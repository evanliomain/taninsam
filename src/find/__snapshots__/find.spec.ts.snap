// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`find function [{ a, b }] |> find(e => e.a === 3) === { a: 3, b: "c" } 1`] = `
Object {
  "a": 3,
  "b": "c",
}
`;

exports[`find function is a pure function 1`] = `
Array [
  Object {
    "a": 1,
    "b": "a",
  },
  Object {
    "a": 2,
    "b": "b",
  },
  Object {
    "a": 3,
    "b": "c",
  },
  Object {
    "a": 4,
    "b": "d",
  },
]
`;
